File: cdm-sources/src/cdm/product/qualification/functions/PerformancePayoutAndFixedPricePayoutOnlyExists.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.product.template.Payout import Payout

__all__ = ['PerformancePayoutAndFixedPricePayoutOnlyExists']


@replaceable
def PerformancePayoutAndFixedPricePayoutOnlyExists(payouts: list[Payout] | None) -> bool:
    """
    At least one PerformancePayout exists, and at least one FixedPricePayout exists, and no other payouts exist.
    
    Parameters 
    ----------
    payouts : Payout
    
    Returns
    -------
    result : boolean
    
    """
    self = inspect.currentframe()
    
    
    result =  ((rosetta_attr_exists(rosetta_resolve_attr(rosetta_resolve_attr(self, "payouts"), "PerformancePayout")) and rosetta_attr_exists(rosetta_resolve_attr(rosetta_resolve_attr(self, "payouts"), "FixedPricePayout"))) and all_elements(map(lambda item: (rosetta_attr_exists(rosetta_resolve_attr(self, "PerformancePayout")) or rosetta_attr_exists(rosetta_resolve_attr(self, "FixedPricePayout"))), rosetta_resolve_attr(self, "payouts")), "=", True))
    
    
    return result

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)

File: cdm-sources/cdm/product/qualification/functions/PerformancePayoutAndFixedPricePayoutOnlyExists.java
package cdm.product.qualification.functions;

import cdm.product.template.FixedPricePayout;
import cdm.product.template.Payout;
import cdm.product.template.PerformancePayout;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperC;
import com.rosetta.model.lib.mapper.MapperS;
import java.util.Collections;
import java.util.List;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(PerformancePayoutAndFixedPricePayoutOnlyExists.PerformancePayoutAndFixedPricePayoutOnlyExistsDefault.class)
public abstract class PerformancePayoutAndFixedPricePayoutOnlyExists implements RosettaFunction {

	/**
	* @param payouts 
	* @return result 
	*/
	public Boolean evaluate(List<? extends Payout> payouts) {
		Boolean result = doEvaluate(payouts);
		
		return result;
	}

	protected abstract Boolean doEvaluate(List<? extends Payout> payouts);

	public static class PerformancePayoutAndFixedPricePayoutOnlyExistsDefault extends PerformancePayoutAndFixedPricePayoutOnlyExists {
		@Override
		protected Boolean doEvaluate(List<? extends Payout> payouts) {
			if (payouts == null) {
				payouts = Collections.emptyList();
			}
			Boolean result = null;
			return assignOutput(result, payouts);
		}
		
		protected Boolean assignOutput(Boolean result, List<? extends Payout> payouts) {
			result = exists(MapperC.<Payout>of(payouts).<PerformancePayout>map("getPerformancePayout", payout -> payout.getPerformancePayout())).and(exists(MapperC.<Payout>of(payouts).<FixedPricePayout>map("getFixedPricePayout", payout -> payout.getFixedPricePayout()))).and(areEqual(MapperC.<Payout>of(payouts)
				.mapItem(item -> exists(item.<PerformancePayout>map("getPerformancePayout", payout -> payout.getPerformancePayout())).or(exists(item.<FixedPricePayout>map("getFixedPricePayout", payout -> payout.getFixedPricePayout()))).asMapper()), MapperS.of(true), CardinalityOperator.All)).get();
			
			return result;
		}
	}
}
